File: installers/ufw.py
# SPDX-FileCopyrightText: Copyright (C) 2025 Akshat Kotpalliwar (alias IntegerAlex) <inquiry.akshatkotpalliwar@gmail.com>
# SPDX-License-Identifier: GPL-3.0-only
"""UFW firewall installer."""

import shutil

from utils.base import BaseInstaller
from utils.types import InstallerResult


class UFWInstaller(BaseInstaller):
    """Install and configure UFW firewall."""
    
    def is_installed(self) -> bool:
        # Check if command exists first to avoid error logs
        if not shutil.which("ufw"):
            return False
        
        try:
            result = self.runner.run(["ufw", "--version"], check=False, capture_output=True)
            return result.returncode == 0
        except Exception:
            return False
    
    def install(self) -> InstallerResult:
        try:
            self.console.print("[blue]ℹ[/blue] Installing UFW")
            
            self.runner.run(
                ["apt", "install", "-y", "ufw"],
                sudo=True,
                description="Installing UFW"
            )
            
            # Configure firewall rules
            self.runner.run(
                ["ufw", "default", "deny", "incoming"],
                sudo=True,
                description="Setting default deny incoming"
            )
            
            self.runner.run(
                ["ufw", "default", "allow", "outgoing"],
                sudo=True,
                description="Setting default allow outgoing"
            )
            
            self.runner.run(
                ["ufw", "allow", "22/tcp"],
                sudo=True,
                description="Allowing SSH"
            )
            
            self.runner.run(
                ["ufw", "allow", "80"],
                sudo=True,
                description="Allowing HTTP"
            )
            
            self.runner.run(
                ["ufw", "allow", "443"],
                sudo=True,
                description="Allowing HTTPS"
            )
            
            self.console.print(
                "[yellow]⚠[/yellow] UFW configured but NOT enabled. Enable manually with: sudo ufw enable"
            )
            
            return InstallerResult(True, "UFW installed and configured (not enabled)")
        except Exception as e:
            return InstallerResult(False, "Failed to install UFW", str(e))

File: src/installers/ufw.ts
import type { CommandRunner } from "../utils/command-runner.js";
import type { SetupLogger } from "../utils/logger.js";
import type { InstallerResult } from "../types.js";
import { blue, yellow } from "../utils/ansi.js";

export function createUFWInstaller(runner: CommandRunner, logger: SetupLogger) {
  function checkInstalled(): boolean {
    try {
      const r = runner.run(["which", "ufw"], { check: false, capture: true });
      if ((r.status ?? 1) !== 0) return false;
      const rr = runner.run(["ufw", "--version"], { check: false, capture: true });
      return (rr.status ?? 1) === 0;
    } catch {
      return false;
    }
  }

  async function doInstall(): Promise<InstallerResult> {
    try {
      process.stdout.write(`${blue("\u2139")} Installing UFW\n`);

      runner.run(["apt", "install", "-y", "ufw"], { sudo: true, description: "Installing UFW" });
      runner.run(["ufw", "default", "deny", "incoming"], { sudo: true, description: "Setting default deny incoming" });
      runner.run(["ufw", "default", "allow", "outgoing"], { sudo: true, description: "Setting default allow outgoing" });
      runner.run(["ufw", "allow", "22/tcp"], { sudo: true, description: "Allowing SSH" });
      runner.run(["ufw", "allow", "80"], { sudo: true, description: "Allowing HTTP" });
      runner.run(["ufw", "allow", "443"], { sudo: true, description: "Allowing HTTPS" });

      process.stdout.write(`${yellow("\u26a0")} UFW configured but NOT enabled. Enable manually with: sudo ufw enable\n`);

      return { success: true, message: "UFW installed and configured (not enabled)" };
    } catch (e) {
      return { success: false, message: "Failed to install UFW", error: String(e) };
    }
  }

  return { is_installed: checkInstalled, install: doInstall };
}
